src/GraphQL/Types/Collections/SchemaTypes.cs

```diff
@@ -646,11 +646,37 @@ private void AddTypeIfNotRegistered(IGraphType type, TypeCollectionContext conte
             var (namedType, namedType2) = type.GetNamedTypes();
             namedType ??= context.ResolveType(namedType2!);
 
-            var foundType = this[namedType.Name];
-            if (foundType == null)
+            var existingType = this[namedType.Name];
+            if (existingType is null)
             {
                 AddType(namedType, context);
             }
+            else
+            {
+                EnsureTypeEquality(existingType, namedType);
+            }
+        }
+
+        private void EnsureTypeEquality(IGraphType existingType, IGraphType newType)
+        {
+            if (ReferenceEquals(existingType, newType))
+            {
+                return;
+            }
+
+            // Ignore scalars
+            if (existingType is ScalarGraphType && newType is ScalarGraphType)
+            {
+                return;
+            }
+
+            if (existingType.GetType() != newType.GetType())
+            {
+                throw new InvalidOperationException($@"Unable to register GraphType '{newType.GetType().FullName}' with the name '{newType.Name}'. The name '{newType.Name}' is already registered to '{existingType.GetType().FullName}'. Check your schema configuration.");
```

Comment:
We have a `GetFriendlyName()` extension method for `Type`s rather than `FullName` which we typically use as a matter of convention.
```suggestion
                throw new InvalidOperationException($@"Unable to register GraphType '{newType.GetType().GetFriendlyName()}' with the name '{newType.Name}'. The name '{newType.Name}' is already registered to '{existingType.GetType().GetFriendlyName()}'. Check your schema configuration.");
```

Reply:
I copied this from another location. Will change it there as well.